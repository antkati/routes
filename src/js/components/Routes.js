import React, { useState, useCallback } from 'react';
import SidebarRoutes from './SidebarRoutes.js';
import MapRoutes from './MapRoutes.js';

export let counter = {value: 0};

let Routes = () => {
    const initialRoutes = {
        points: {},
        columns: {
            'column-1': {
                id: 'column-1',
                pointsIds: [],
            }
        },
        columnOrder: ['column-1'],
    }

    const [routes, setRoutes] = useState(initialRoutes);
    const [mapCenter, setMapCenter] = useState([56.02279830, 92.89742900]);


    //add coords
    const handleInput = useCallback(
        (e) => {
            if (e.key !== 'Enter') return;

            const inputValue = e.target.value;
            if (inputValue === '') return;

            const id = counter.value;

            const column = routes.columns['column-1'];
            const newPointsIds = Array.from(column.pointsIds);
            newPointsIds.push(`point-${id}`);
            const newColumn = {
                ...column,
                pointsIds: newPointsIds
            }

            const newRoutes = {
                ...routes,
                points: {
                    ...routes.points,
                    [`point-${id}`]: {
                        id: `point-${id}`,
                        content: inputValue,
                        coords: mapCenter,
                    },
                },
                columns: {
                    ...routes.columns,
                    ['column-1']: newColumn,
                }
            };

            setRoutes(newRoutes);

            e.target.value = '';
            counter.value++;
        },
        [routes, mapCenter]
    )

    //change coords of placemark
    const handlePlacemarkDrag = useCallback(
        (e) => {
            const target = e.get('target');
            const id = target.properties.get('id');
            const currentCoords = target.geometry.getCoordinates();

            const point = routes.points[id];
            const newPoint = {
                ...point,
                coords: currentCoords,
            }

            const newRoutes = {
                ...routes,
                points: {
                    ...routes.points,
                    [id]: newPoint,
                },
            };

            setRoutes(newRoutes);
        },
        [routes]
    )

    //change mapCenter
    const handleMapBoundsChange = useCallback(
        (e) => {
            const newCenter = e.get('newCenter');
            setMapCenter(newCenter);
        },
        [mapCenter]
    )

    //delete placemark
    const handleSidebarItemDelete = useCallback(
        (e) => {
            let target = e.target;
            if (!target.classList.contains('close-item')) return;

            while (target) {
                if (target.tagName == 'LI') break;
                target = target.parentNode;
            }

            if (!target) return;

            if (!target.hasAttribute('data-id') || !target.hasAttribute('data-index')) return;

            const id = target.dataset.id;
            const index = target.dataset.index;

            const column = routes.columns['column-1'];
            const newPointsIds = Array.from(column.pointsIds);
            newPointsIds.splice(index, 1);
            const newColumn = {
                ...column,
                pointsIds: newPointsIds
            }

            delete routes.points[id];

            const newRoutes = {
                ...routes,
                columns: {
                    ...routes.columns,
                    ['column-1']: newColumn,
                },
            }

            setRoutes(newRoutes);
        },
        [routes]
    )

    //change order of placemarks
    const handleSidebarItemDrag = useCallback(
        (result) => {
            const { source, destination, draggableId } = result;
            if (!destination) return;
            if (destination.droppableId === source.droppableId && destination.index === source.index) return;

            const column = routes.columns['column-1'];
            const newPointsIds = Array.from(column.pointsIds);

            newPointsIds.splice(source.index, 1);
            newPointsIds.splice(destination.index, 0, draggableId);

            const newColumn = {
                ...column,
                pointsIds: newPointsIds
            }

            const newRoutes = {
                ...routes,
                columns: {
                    ...routes.columns,
                    ['column-1']: newColumn,
                }
            }

            setRoutes(newRoutes);
        },
        [routes]
    )

    const propsSidebarRoutes = {
        routes,
        handleInput,
        handleSidebarItemDelete,
        handleSidebarItemDrag,
    }

    const propsMapRoutes = {
        routes,
        mapCenter,
        handleMapBoundsChange,
        handlePlacemarkDrag,
    }

    return (
        <React.Fragment>
            <SidebarRoutes {...propsSidebarRoutes} />
            <MapRoutes {...propsMapRoutes} />
        </React.Fragment>
    )
}

Routes = React.memo(Routes);

export default Routes;