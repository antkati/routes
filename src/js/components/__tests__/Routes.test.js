import React from 'react';
import { shallow, mount } from 'enzyme';
import Routes from '../Routes.js';
import { counter } from '../Routes.js';

describe('Routes', () => {
    let renderedComponent;

    function mountComponent() {
        counter.value = 0;
        renderedComponent = mount(<Routes />);
    }

    test('Routes component renders correctly', () => {
        counter.value = 0;
        const renderedComponent = shallow(<Routes />);
        expect(renderedComponent).toMatchSnapshot();
    });


    test('Routes component renders with initial state', () => {
        mountComponent();

        const lisLength = renderedComponent.find('.map-sidebar__list__item').length;
        expect(lisLength).toBe(0);
    });


    test('InputHandle does not add li with empty value', () => {
        mountComponent();

        const input = renderedComponent.find('.map-sidebar__input');
        
        input.simulate('keydown', {key: 'Enter', target: { value: '' }});
        let lisLength = renderedComponent.find('.map-sidebar__list__item').length;
        expect(lisLength).toBe(0);

        input.simulate('keydown', {key: 'Enter', target: { value: 'Точка маршрута 1' }});
        lisLength = renderedComponent.find('.map-sidebar__list__item').length;
        expect(lisLength).toBe(1);
    });  


    test('InputHandle adds li with Enter key only', () => {
        mountComponent();

        const input = renderedComponent.find('.map-sidebar__input');

        input.simulate('keydown', {key: 'Tab', target: { value: 'Точка маршрута 1' }});
        let lisLength = renderedComponent.find('.map-sidebar__list__item').length;
        expect(lisLength).toBe(0);

        input.simulate('keydown', {key: 'Control', target: { value: 'Точка маршрута 1' }});
        lisLength = renderedComponent.find('.map-sidebar__list__item').length;
        expect(lisLength).toBe(0);

        input.simulate('keydown', {key: 'Enter', target: { value: 'Точка маршрута 1' }});
        lisLength = renderedComponent.find('.map-sidebar__list__item').length;
        expect(lisLength).toBe(1);
    });

    
    describe('InputHandle adds li right', () => {
        mountComponent();

        const input = renderedComponent.find('.map-sidebar__input');

        describe('first li was added', () => {
            const inputValue = 'Точка маршрута 1';
            input.simulate('keydown', {key: 'Enter', target: { value: inputValue }});
            const lis = renderedComponent.find('.map-sidebar__list__item');
            const li = lis.first();

            test('number of lis increased by one', () => {
                expect(lis.length).toBe(1);
            });

            test('li has valid id', () => {
                expect(li.prop('data-id')).toBe('point-0');
            });

            test('li has been added to the end of list', () => {
                expect(li.prop('data-index')).toBe(lis.length - 1);
            });

            test('li text is equal to input value', () => {
                const liText = li.find('span').first().text();
                expect(liText).toBe(inputValue);
            });
        });


        describe('second li was added', () => {
            const inputValue = 'Точка маршрута 2';
            input.simulate('keydown', {key: 'Enter', target: { value: inputValue }});
            const lis = renderedComponent.find('.map-sidebar__list__item');
            const li = lis.at(1);

            test('number of lis increased by one', () => {
                expect(lis.length).toBe(2);
            });

            test('li has valid id', () => {
                expect(li.prop('data-id')).toBe('point-1');
            });

            test('li has been added to the end of list ', () => {
                expect(li.prop('data-index')).toBe(lis.length - 1);
            });

            test('li text is equal to input value', () => {
                const liText = li.find('span').first().text();
                expect(liText).toBe(inputValue);
            });
        })

    });


    test('handleSidebarItemDelete do not work with ul nested elements except .close-item', () => {
        mountComponent();

        const input = renderedComponent.find('.map-sidebar__input');
        input.simulate('keydown', {key: 'Enter', target: { value: 'Точка маршрута 1' }}); 
        
        let lis = renderedComponent.find('.map-sidebar__list__item');
        const li = lis.first();
        li.simulate('click');
        lis = renderedComponent.find('.map-sidebar__list__item');
        expect(lis.length).toBe(1);

        const content = li.find('span').first();
        content.simulate('click');
        lis = renderedComponent.find('.map-sidebar__list__item');
        expect(lis.length).toBe(1);

        const closeItem = li.find('.close-item');
        closeItem.simulate('click');
        lis = renderedComponent.find('.map-sidebar__list__item');
        expect(lis.length).toBe(0);
    });


    describe('handleSidebarItemDelete removes li right', () => {
        mountComponent();

        const input = renderedComponent.find('.map-sidebar__input');
        const values = ['точка маршрута 1', 'точка маршрута 2', 'точка маршрута 3'];
        values.forEach(value => {
            input.simulate('keydown', {key: 'Enter', target: { value: value }});         
        });

        let lis = renderedComponent.find('.map-sidebar__list__item');

        const li = lis.at(1);
        const id = li.prop('data-id');
        const closeItem = li.find('.close-item');
        closeItem.simulate('click');

        lis = renderedComponent.find('.map-sidebar__list__item');

        test('number of lis decreased by one', () => {
            expect(lis.length).toBe(2);
        });

        test('deleted li has valid id', () => {
            expect(lis.exists(`li[data-id='${id}']`)).toBeFalsy();  
        });

        test('remaining lis changed their data-index according to new order', () => {
            lis.forEach((li, index) => {
                expect(li.prop('data-index')).toBe(index);
            })
        });
    });    


    test('lis have unique id after repeated adding and removing', () => {
        mountComponent();

        const input = renderedComponent.find('.map-sidebar__input');
        const values = ['точка маршрута 1', 'точка маршрута 2', 'точка маршрута 3'];
        values.forEach(value => {
            input.simulate('keydown', {key: 'Enter', target: { value: value }});         
        });       
        
        let lis = renderedComponent.find('.map-sidebar__list__item');

        const li = lis.at(1);
        const closeItem = li.find('.close-item');
        closeItem.simulate('click');

        input.simulate('keydown', {key: 'Enter', target: { value: 'еще одна точка маршрута' }});
        lis = renderedComponent.find('.map-sidebar__list__item');
        
        lis.forEach((li) => {
            const id = li.prop('data-id');
            expect(lis.find(`li[data-id='${id}']`).length).toBe(1);  
        })
    });
})

