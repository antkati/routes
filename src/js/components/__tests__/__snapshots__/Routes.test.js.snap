// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routes Routes component renders correctly 1`] = `
<Fragment>
  <Memo(SidebarRoutes)
    handleInput={[Function]}
    handleSidebarItemDelete={[Function]}
    handleSidebarItemDrag={[Function]}
    routes={
      Object {
        "columnOrder": Array [
          "column-1",
        ],
        "columns": Object {
          "column-1": Object {
            "id": "column-1",
            "pointsIds": Array [],
          },
        },
        "points": Object {},
      }
    }
  />
  <MapRoutes
    handleMapBoundsChange={[Function]}
    handlePlacemarkDrag={[Function]}
    mapCenter={
      Array [
        56.0227983,
        92.897429,
      ]
    }
    routes={
      Object {
        "columnOrder": Array [
          "column-1",
        ],
        "columns": Object {
          "column-1": Object {
            "id": "column-1",
            "pointsIds": Array [],
          },
        },
        "points": Object {},
      }
    }
  />
</Fragment>
`;
